x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/persistence/AnomalyDetectorsIndex.java

```diff
@@ -89,64 +107,105 @@ public static String configIndexName() {
     }
 
     /**
-     * Create the .ml-state index (if necessary)
-     * Create the .ml-state-write alias for the .ml-state index (if necessary)
+     * Creates the .ml-state-000001 index (if necessary)
+     * Creates the .ml-state-write alias for the .ml-state-000001 index (if necessary)
      */
     public static void createStateIndexAndAliasIfNecessary(Client client, ClusterState state, final ActionListener<Boolean> finalListener) {
-
-        if (state.getMetaData().getAliasAndIndexLookup().containsKey(jobStateIndexWriteAlias())) {
-            finalListener.onResponse(false);
-            return;
+        IndexNameExpressionResolver indexNameExpressionResolver = new IndexNameExpressionResolver();
+        String[] stateIndices =
+            indexNameExpressionResolver.concreteIndexNames(state, IndicesOptions.lenientExpandOpen(), jobStateIndexPattern());
+        Optional<IndexMetaData> indexPointedByCurrentWriteAlias = state.getMetaData().hasAlias(".ml-state-write")
+            ? state.getMetaData().getAliasAndIndexLookup().get(".ml-state-write").getIndices().stream().findFirst()
+            : Optional.empty();
+        String legacyJobStateIndex = AnomalyDetectorsIndexFields.STATE_INDEX_PREFIX;
+
+        if (stateIndices.length == 0) {
+            if (indexPointedByCurrentWriteAlias.isEmpty()) {
+                createInitialStateIndex(client, true, finalListener);
+                return;
+            }
+            logger.error(
+                "There are no indices matching '.ml-state*' pattern but '.ml-state-write' alias points at [{}]. "
+                + "This should never happen.",
+                indexPointedByCurrentWriteAlias.get());
+        } else if (stateIndices.length == 1 && stateIndices[0].equals(legacyJobStateIndex)) {
+            if (indexPointedByCurrentWriteAlias.isEmpty()) {
+                createInitialStateIndex(client, true, finalListener);
+                return;
+            }
+            if (indexPointedByCurrentWriteAlias.get().getIndex().getName().equals(legacyJobStateIndex)) {
+                createInitialStateIndex(
+                    client,
+                    false,
+                    ActionListener.wrap(
+                        unused -> updateStateWriteAlias(client, legacyJobStateIndex, initialJobStateIndex(), finalListener),
+                        finalListener::onFailure)
+                );
+                return;
+            }
+            logger.error(
+                "There is exactly one index (i.e. '.ml-state') matching '.ml-state*' pattern but '.ml-state-write' alias points at [{}]. "
+                + "This should never happen.",
+                indexPointedByCurrentWriteAlias.get());
+        } else {
+            if (indexPointedByCurrentWriteAlias.isEmpty()) {
+                Arrays.sort(stateIndices, Collections.reverseOrder());
```

Comment:
There's an edge case here, which is that it's sorting on the _concrete_ index names, and those could have been modified during a major version upgrade.

For example, if `.ml-state` was first created in 5.5 then in 6.8 the migration assistant would have reindexed it to `.reindexed-6-ml-state` with an alias of `.ml-state` (in addition to the `.ml-state-write` alias).  Sorting in reverse would put `. reindexed-6-ml-state` before `.ml-state-000042`.

It would be safer to create a custom comparator that extracts the last 6 digits from the name for concrete names that end with a dash followed by 6 digits, and `000000` otherwise.

Reply:
I rebased this PR on the `.ml-state` renaming PR and fixed the tests.
PTAL